In [240]:
#Download libraries
import pandas as pd
import altair as alt
import numpy as np
import matplotlib.pyplot as plt

In [241]:
#Download the data
df = pd.read_csv('../data/train_data.zip')
df.head()

external_id  month  year  monthly_number_of_sessions  \
0     1807558      9  2018                           0   
1     1806729      9  2018                           0   
2  MR00096403      9  2018                           0   
3     1806350      9  2018                           0   
4  FM00169477      9  2018                           0   

   monthly_unique_sessions  monthly_repeated_sessions  \
0                        0                          0   
1                        0                          0   
2                        0                          0   
3                        0                          0   
4                        0                          0   

   monthly_avg_length_of_session  monthly_avg_light_activity  \
0                            0.0                         0.0   
1                            0.0                         0.0   
2                            0.0                         0.0   
3                            0.0                         0.0   
4                            0.0                         0.0   

   monthly_avg_moderate_activity  monthly_avg_vigorous_activity  ...  \
0                            0.0                            0.0  ...   
1                            0.0                            0.0  ...   
2                            0.0                            0.0  ...   
3                            0.0                            0.0  ...   
4                            0.0                            0.0  ...   

   avg_wind_9_10  avg_wind_10_11  avg_wind_11_12  avg_wind_12_above  \
0            0.0             0.0             0.0                0.0   
1            0.0             0.0             0.0                0.0   
2            0.0             0.0             0.0                0.0   
3            0.0             0.0             0.0                0.0   
4            0.0             0.0             0.0                0.0   

   perfect_days  unacast_session_count     hpi  \
0           0.0                   37.0  418.67   
1           0.0                   63.0  710.00   
2           1.0                   15.0  241.24   
3           2.0                   45.0  468.51   
4           0.0                   83.0  326.05   

   state_and_local_amount_per_capita  state_amount_per_capita  \
0                           0.116368                 0.025516   
1                           0.175774                 0.020170   
2                           0.112038                 0.006893   
3                           0.120088                 0.003619   
4                           0.132935                 0.025298   

   local_amount_per_capita  
0                 0.090852  
1                 0.155604  
2                 0.105145  
3                 0.116469  
4                 0.107637  

[5 rows x 861 columns]

## Study of the NaNs is the temperature data

In [242]:
df[df['avg_wind_12_above'].isnull()]

external_id  month  year  monthly_number_of_sessions  \
556     1901754     11  2018                           0   

     monthly_unique_sessions  monthly_repeated_sessions  \
556                        0                          0   

     monthly_avg_length_of_session  monthly_avg_light_activity  \
556                            0.0                         0.0   

     monthly_avg_moderate_activity  monthly_avg_vigorous_activity  ...  \
556                            0.0                            0.0  ...   

     avg_wind_9_10  avg_wind_10_11  avg_wind_11_12  avg_wind_12_above  \
556            NaN             NaN             NaN                NaN   

     perfect_days  unacast_session_count     hpi  \
556           NaN                    8.0  470.64   

     state_and_local_amount_per_capita  state_amount_per_capita  \
556                           0.173926                 0.019939   

     local_amount_per_capita  
556                 0.153987  

[1 rows x 861 columns]

There is only one observation with missing values in the temperature data. Let's study the playground that is the closest to this playground (regarding to the latitude and the longitude), and see if they have similar values for the other months.

In [114]:
my_dist = df.iloc[556][['longitude', 'latitude']]
my_dist[0]

-156.388651

In [116]:
dist = np.square(df['longitude'] - my_dist[0] ) +  np.square(df['latitude'] - my_dist[1])

In [125]:
dist.sort_values().head(21)

20458     0.000000
35507     0.000000
35495     0.000000
35477     0.000000
35395     0.000000
30349     0.000000
35390     0.000000
45631     0.000000
10481     0.000000
5621      0.000000
30400     0.000000
5587      0.000000
20510     0.000000
25446     0.000000
503       0.000000
5537      0.000000
40548     0.000000
546       0.000000
25433     0.000000
556       0.000000
4205     62.768606
dtype: float64

Check that the playgrounds have very close longitude and latitude.

In [127]:
df.iloc[4205][['longitude', 'latitude']]

longitude    -149.87
latitude     61.1987
Name: 4205, dtype: object

In [128]:
my_dist

longitude   -156.389
latitude     65.7019
Name: 556, dtype: object

In [142]:
temp = df[df['external_id'].isin(['MR00111851', '1901754'])].sort_values(['year', 'month'])

In [144]:
temp.loc[:,'weather_clear':'avg_wind_12_above']

weather_clear  weather_rain  weather_fog  weather_snow  \
19152           18.0           0.0          0.0           8.0   
20458           10.0           0.0          2.0          18.0   
4205            21.0           0.0          0.0           6.0   
30349            5.0           2.0          1.0          26.0   
14218           28.0           0.0          0.0           0.0   
35477           22.0           7.0          1.0           6.0   
10481           10.0          13.0          2.0           3.0   
34091           24.0           0.0          0.0           0.0   
19164           25.0           0.0          0.0           0.0   
45631           18.0           6.0          1.0           0.0   
30400           19.0           9.0          2.0           0.0   
49372           25.0           0.0          0.0           0.0   
9251            12.0           0.0          0.0           0.0   
25433            8.0          13.0          0.0           0.0   
14192           27.0           0.0          0.0           0.0   
40548           10.0           8.0          5.0           0.0   
503             12.0           8.0          1.0           5.0   
29024           16.0           0.0          0.0           1.0   
556              NaN           NaN          NaN           NaN   
19178           21.0           0.0          0.0           3.0   
546              7.0           0.0          8.0          24.0   
19173            9.0           0.0          0.0          16.0   
9271            25.0           0.0          0.0           2.0   
35507           10.0           1.0          7.0          19.0   
5537             6.0           3.0          6.0          21.0   
39213           20.0           0.0          0.0           7.0   
5587            12.0           4.0          5.0          18.0   
29088           29.0           0.0          0.0           1.0   
9265            21.0           0.0          0.0           6.0   
25446           16.0           7.0          3.0          12.0   
14161           25.0           0.0          0.0           0.0   
35395           12.0          18.0          1.0           1.0   
35390           22.0           8.0          1.0           0.0   
49308           30.0           0.0          0.0           0.0   
34133           27.0           0.0          0.0           0.0   
35495           19.0          11.0          2.0           0.0   
20510           18.0          11.0          2.0           0.0   
49379           30.0           0.0          0.0           0.0   
4276            16.0           0.0          0.0           0.0   
5621             0.0           4.0          0.0           0.0   

       weather_thunder  weather_cloudy  weather_other  temp_avg_35_below  \
19152              0.0             7.0            0.0               28.0   
20458              0.0            16.0            0.0               28.0   
4205               0.0             8.0            0.0               29.0   
30349              0.0            25.0            0.0               31.0   
14218              0.0             2.0            0.0                5.0   
35477              0.0             7.0            0.0               24.0   
10481              0.0            13.0            0.0                3.0   
34091              0.0             7.0            0.0                0.0   
19164              0.0             5.0            0.0                0.0   
45631              0.0             6.0            0.0                0.0   
30400              0.0             8.0            0.0                0.0   
49372              0.0             7.0            0.0                0.0   
9251               0.0            20.0            0.0                0.0   
25433              0.0            13.0            0.0                0.0   
14192              0.0             4.0            0.0                0.0   
40548              0.0             4.0            0.0                0.0   
503                0.0             9.0           

The values don't seem to be similar between the two playgrounds. I would just impute the missing values with the means of the values for the same playground, for the previous and the following month.

## Find states with missing values for the 'political' data

In [152]:
df[df['Democrats_12_Votes'].isnull()]['state'].unique()

array(['Alaska'], dtype=object)

In [154]:
df[df['Republicans_2016'].isnull()]['state'].unique()

array(['Alaska'], dtype=object)

In [158]:
df[df['Green_2016'].isnull()]['state'].unique()

array(['Georgia', 'North Carolina', 'Nevada', 'Indiana', 'South Dakota',
       'Alaska', 'Oklahoma'], dtype=object)

## Study of the `unacast_session_count` 

Let's study the playgrounds that only have missing values until a given month.

In [209]:
#my_dict will contains the id of the playgrounds with missing values until a given month and a given year.
my_dict = {'month':[temp[0][0]], 'year':[temp[0][1]], 'playgrounds':[list(df[df['unacast_session_count'].isnull()].query('year==2018 & month==2')['external_id'].values)]}
for i in range(len(temp)-1):
    list_play = list(df[df['unacast_session_count'].isnull()].query('year==' + str(temp[i+1][1]) + ' & month==' + str(temp[i+1][0]))['external_id'].values)
    #Check that the id playground is in the list of playgrounds with missing target for the month under study, and all the privious months
    name_play = list(set(list_play) & set(my_dict['playgrounds'][-1]))

    my_dict['month'].append(temp[i+1][0])
    my_dict['year'].append(temp[i+1][1])
    my_dict['playgrounds'].append(name_play)


In [217]:
playgrounds = pd.DataFrame(my_dict)
#Add the count of playgrounds that only have missing values until the given month
playgrounds['count'] = list(map(lambda x : len(x), playgrounds['playgrounds']))

In [218]:
playgrounds

month  year                                        playgrounds  count
0       2  2018  [MR00096956, 1805327, FM00177197, e652e9a0-320...    124
1       3  2018  [MR00114723, FM00171867, 5ba4de98-f4cc-4c60-94...     43
2       4  2018  [FM00171867, 1809358, 1809195, MR00110941, MR0...     17
3       5  2018  [FM00171867, 1809130, MR00105990, 1805183, MR0...      7
4       6  2018   [MR00101262, FM00171867, MR00105990, MR00095127]      4
5       7  2018   [MR00101262, FM00171867, MR00105990, MR00095127]      4
6       8  2018   [MR00101262, FM00171867, MR00105990, MR00095127]      4
7       9  2018   [MR00101262, FM00171867, MR00105990, MR00095127]      4
8      10  2018   [MR00101262, FM00171867, MR00105990, MR00095127]      4
9      11  2018               [MR00101262, MR00105990, MR00095127]      3
10     12  2018                           [MR00101262, MR00105990]      2
11      1  2019                                       [MR00101262]      1
12      2  2019                                       [MR00101262]      1
13      3  2019                                       [MR00101262]      1
14      4  2019                                       [MR00101262]      1
15      5  2019                                                 []      0
16      6  2019                                                 []      0
17      7  2019                                                 []      0
18      8  2019                                                 []      0

The dataframe above has to be understood in the following way : if we take for example the row with the index 7, it means that the playgrounds whose external_id belongs in [MR00101262, FM00171867, MR00105990, MR00095127] only have missing target value until September 2018 (so for those playgrounds, the `unicast_session_count` is NaN for all the months between February 2018 and September 2018).

In [220]:
playgrounds['count'].sum()

220

Let's delete the corresponding rows from the original dataset (as we may suspect that those values are NaNs because the playground didn't exist yet).

In [234]:
my_df = df.copy()
for i in range(playgrounds.shape[0]):
    year = playgrounds['year'][i]
    month = playgrounds['month'][i]
    ids = playgrounds['playgrounds'][i]
    my_df = my_df[((my_df.year != year) | (my_df.month != month) | (my_df.external_id.isin(ids) == False))]

In [235]:
my_df

external_id  month  year  monthly_number_of_sessions  \
0                1807558      9  2018                           0   
1                1806729      9  2018                           0   
2             MR00096403      9  2018                           0   
3                1806350      9  2018                           0   
4             FM00169477      9  2018                           0   
...                  ...    ...   ...                         ...   
50115         MR00111485      2  2018                           0   
50116  Kyle Rec Kenner 4      2  2018                           0   
50117            1901693      2  2018                           0   
50118         FM00166877      2  2018                           0   
50119         MR00096958      2  2018                           0   

       monthly_unique_sessions  monthly_repeated_sessions  \
0                            0                          0   
1                            0                          0   
2                            0                          0   
3                            0                          0   
4                            0                          0   
...                        ...                        ...   
50115                        0                          0   
50116                        0                          0   
50117                        0                          0   
50118                        0                          0   
50119                        0                          0   

       monthly_avg_length_of_session  monthly_avg_light_activity  \
0                                0.0                         0.0   
1                                0.0                         0.0   
2                                0.0                         0.0   
3                                0.0                         0.0   
4                                0.0                         0.0   
...                              ...                         ...   
50115                            0.0                         0.0   
50116                            0.0                         0.0   
50117                            0.0                         0.0   
50118                            0.0                         0.0   
50119                            0.0                         0.0   

       monthly_avg_moderate_activity  monthly_avg_vigorous_activity  ...  \
0                                0.0                            0.0  ...   
1                                0.0                            0.0  ...   
2                                0.0                            0.0  ...   
3                                0.0                            0.0  ...   
4                                0.0                            0.0  ...   
...                              ...                            ...  ...   
50115                            0.0                            0.0  ...   
50116                            0.0                            0.0  ...   
50117                            0.0                            0.0  ...   
50118                            0.0                            0.0  ...   
50119                            0.0                            0.0  ...   

       avg_wind_9_10  avg_wind_10_11  avg_wind_11_12  avg_wind_12_above  \
0                0.0             0.0             0.0                0.0   
1                0.0             0.0             0.0                0.0   
2                0.0             0.0             0.0                0.0   
3                0.0             0.0             0.0                0.0   
4                0.0             0.0             0.0                0.0   
...              ...             ...             ...                ...   
50115            0.0             0.0             0.0                0.0   
50116            0.0             0.0             0.0                0.0   
50117            0.0             0.0          

In [243]:
#Check if the dimensions of my_df is right
df.shape[0] - playgrounds['count'].sum() == my_df.shape[0]

True

In [238]:
my_df[my_df['unacast_session_count'].isnull()]['external_id'].value_counts()

e5bdd7bc-325a-4f88-ad22-a3a820118b68    15
MR00110941                              15
MR00117367                              13
1805183                                  7
MR00118229                               7
                                        ..
MR00108479                               1
1810710                                  1
1810866                                  1
FM00168120                               1
1806104                                  1
Name: external_id, Length: 164, dtype: int64

In [239]:
df[df['unacast_session_count'].isnull()]['external_id'].value_counts()

MR00101262                              19
MR00105990                              18
MR00110941                              18
MR00117367                              16
e5bdd7bc-325a-4f88-ad22-a3a820118b68    15
                                        ..
MR00115778                               1
f83e4ef8-50eb-11e9-8647-d663bd873d93     1
FM00162998                               1
MR00097422                               1
4a317c04-d618-45d3-adde-6e643505b7d9     1
Name: external_id, Length: 222, dtype: int64

We can see that once we delete the rows that correspond to non-existing playgrounds, we only have 164 playgrounds with NaNs values for the `unacast_session_count`.